fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/RecordCursorTest.java

```diff
@@ -302,13 +303,67 @@ public void filterTest() {
             }
         }, 1);
         newInts = cursor.asList().join();
-        assertTrue(cursor instanceof MapCursor, "Creating an async filter should create a map cursor");
-        assertEquals(Arrays.asList(2,4,6), newInts);
+        assertTrue(cursor instanceof MapResultCursor, "Creating an async filter should create a map cursor");
+        assertEquals(Arrays.asList(2, 4, 6), newInts);
     }
 
     @Test
-    public void firstTest() throws Exception {
-        List<Integer> ints = Arrays.asList(1,2,3,4);
+    void mapContinuationsTest() {
+        final List<Integer> ints = Arrays.asList(1, 2, 3, 4, 5, 6);
+        final ByteString prefix = ByteString.copyFromUtf8("prefix+");
+        RecordCursor<Integer> cursor = RecordCursor.fromList(ints)
+                .mapContinuation(continuation -> {
+                    if (continuation.isEnd()) {
+                        return RecordCursorEndContinuation.END;
+                    }
+
+                    return new RecordCursorContinuation() {
+                        @Nonnull
+                        @Override
+                        public ByteString toByteString() {
+                            return prefix.concat(continuation.toByteString());
+                        }
+
+                        @Nullable
+                        @Override
+                        public byte[] toBytes() {
+                            return toByteString().toByteArray();
+                        }
+
+                        @Override
+                        public boolean isEnd() {
+                            return false;
+                        }
+                    };
+                });
+
+        List<Integer> soFar = new ArrayList<>();
+        RecordCursorResult<Integer> result;
+        do {
+            result = cursor.getNext();
+            if (result.getContinuation().isEnd()) {
+                assertEquals(ByteString.EMPTY, result.getContinuation().toByteString());
+                assertNull(result.getContinuation().toBytes());
+            } else {
+                soFar.add(result.get());
+
+                // Modified continuation should begin with the prefix
+                assertTrue(result.getContinuation().toByteString().startsWith(prefix));
+
+                // Stripping away the prefix and resuming the cursor should produce the rest of the list
+                byte[] continuation = result.getContinuation().toBytes();
+                assertNotNull(continuation);
+                RecordCursor<Integer> tailCursor = RecordCursor.fromList(ints, Arrays.copyOfRange(continuation, prefix.size(), continuation.length));
```

Comment:
I do wonder whether this is a problematic api, since:
```
RecordCursor.fromList(ints, continuation).mapContinuation(continuation -> { ... });
```
could be surprising, since the continuation does not line up.
This is similar to `flatMapPipelined`, and perhaps this should be a static method like that, but doing so would require providing a `ContinuationConverter` class (or two lambdas) that can do bidirectional conversion. 

Potentially more importantly, the code as you have it might make it harder to decouple the code creating the continuation from the one mapping the continuation. Maybe, it's not a real issue, I'm having trouble thinking of a situation where there isn't something that passes the continuation into the code creating cursor, and also calls mapContinuation.

Reply:
Yeah, the way that it's the caller's responsibility to manipulate the continuation does feel like it could be a bit of a foot-gun. The alternative using a static method might look something like this:

```java
    interface ContinuationConvertor {
        @Nullable
        byte[] unwrapContinuation(@Nullable byte[] continuation);

        RecordCursorContinuation wrapContinuation(@Nonnull RecordCursorContinuation continuation);
    }

    static <T> RecordCursor<T> mapContinuation(@Nonnull Function<byte[], RecordCursor<T>> cursorFunction, @Nonnull ContinuationConvertor convertor, @Nullable byte[] continuation) {
        byte[] innerContinuation = convertor.unwrapContinuation(continuation);
        return cursorFunction.apply(innerContinuation)
                .mapContinuation(convertor::wrapContinuation);
    }
```

(Note that it uses the current `.mapContinuation` value in its implementation, but that could theoretically be rewritten.)

Which...maybe isn't too bad, but it doesn't feel great.